### PR TITLE
WIP : Comp to vulkan

### PR DESCRIPTION
**Pull request by XingHongChenIntel**
draft comp to vulkan pass base on comp to llvm pass.

### CODE REVIEW

File: pmlc/conversion/comp_to_llvm/BUILD [hunk 1]
```diff
@@ -24,24 +24,26 @@ plaidml_cc_library(
     name = "comp_to_llvm",
     srcs = [
         "common.cc",
-        "comp_to_ocl.cc",
+#        "comp_to_ocl.cc",
```
Comment (YangleiZouIntel) [inline pmlc/conversion/comp_to_llvm/BUILD]: I think we need a separate library for comp to vulkan.

File: pmlc/target/intel_gen_ocl_spirv/pipeline.cc [hunk 2]
```diff
@@ -98,8 +98,9 @@ void pipelineBuilder(OpPassManager &pm) {
   pm.addPass(conversion::gpu::createGpuKernelOutliningPass());
 
   // Convert GPU to comp.
+  // TODO figure out the memorySpace for Vulkan
   pm.addPass(pmlc::conversion::gpu_to_comp::createConvertGpuToCompPass(
-      comp::ExecEnvRuntime::OpenCL, /*memorySpace=*/11));
```
Comment (YangleiZouIntel) [inline pmlc/target/intel_gen_ocl_spirv/pipeline.cc:125]: I think you can move this passes to intel_gen, I think we will rename it to something like inte_gen_vulkan

File: pmlc/conversion/comp_to_llvm/comp_to_vulkan.cc [hunk 3]
```diff
@@ -0,0 +1,472 @@
...
+  // llvm function and variadic arguments.
+  for (mlir::Value event : operands.slice(1)) {
+    rewriter.create<LLVM::CallOp>(loc, mlir::ArrayRef<mlir::Type>{},
+                                  rewriter.getSymbolRefAttr(kAddVulkanLaunchActionToSchedule),
+                                  mlir::ArrayRef<mlir::Value>{operands[0]});
+  }
+  return mlir::success();
+}
+
+std::unique_ptr<mlir::Pass> createConvertCompToOclPass() {
```
Comment (YangleiZouIntel) [inline pmlc/conversion/comp_to_llvm/comp_to_vulkan.cc]: rename pass

File: pmlc/conversion/comp_to_llvm/comp_to_vulkan.cc [hunk 4]
```diff
@@ -0,0 +1,472 @@
...
+  mlir::Type mlirIndexType = builder.getIndexType();
+
+  builder.create<LLVM::LLVMFuncOp>(
+      loc, kInitVulkan,
+      LLVM::LLVMType::getFunctionTy(llvmInt8Ptr, {llvmInt8Ptr},
+          /*isVarArg=*/false));
+
+  builder.create<mlir::FuncOp>(
+      loc, kCreateVulkanLaunchKernelAction,
+      mlir::FunctionType::get(
```
Comment (YangleiZouIntel) [inline pmlc/conversion/comp_to_llvm/comp_to_vulkan.cc]: I think it is better to use llvmInt32 Type than mlir type, so we can use LLVM::LLVMFuncOp insteadof mlir::FuncOp.
Please see
https://github.com/plaidml/plaidml/pull/1356/files#diff-b5c69b0848345bd7a7b84e4e3d744e49R490
https://github.com/plaidml/plaidml/pull/1356/files#diff-b5c69b0848345bd7a7b84e4e3d744e49R588
https://github.com/plaidml/plaidml/pull/1356/files#diff-bd02d185b766dd5f7de33b0148b25089L116

File: pmlc/conversion/comp_to_llvm/comp_to_vulkan.cc [hunk 5]
```diff
@@ -0,0 +1,472 @@
...
+        return llvmInt8Ptr;
+      });
+  patterns.insert<ConvertToInitVulkan>(kInitVulkan, typeConverter, context);
+  patterns.insert<ConvertToDeinitVulkan>(kDeinitVulkan, typeConverter, context);
+  patterns.insert<ConvertToCommandBufferSubmit>(kRun, typeConverter, context);
+
+  patterns.insert<ConvertScheduleFunc>(modulesMap, typeConverter, context);
+  patterns.insert<ConvertAlloc>(typeConverter, context);
+
+  // TODO modify those to vulkan api
```
Comment (YangleiZouIntel) [inline pmlc/conversion/comp_to_llvm/comp_to_vulkan.cc]: Rob should soon have a new revision for [RFC:-Network-Pre-Initializationl](https://github.com/plaidml/plaidml/wiki/RFC:-Network-Pre-Initializationl), we should refer this design.

File: pmlc/conversion/comp_to_llvm/BUILD (hunk 1 above)
Comment (XingHongChenIntel) [reply pmlc/conversion/comp_to_llvm/BUILD]: fixed

File: pmlc/target/intel_gen_ocl_spirv/pipeline.cc (hunk 2 above)
Comment (XingHongChenIntel) [reply pmlc/target/intel_gen_ocl_spirv/pipeline.cc:125]: fixed

File: pmlc/conversion/comp_to_llvm/comp_to_vulkan.cc (hunk 3 above)
Comment (XingHongChenIntel) [reply pmlc/conversion/comp_to_llvm/comp_to_vulkan.cc]: fixed

**Comment by YangleiZouIntel** [comment]
added some of the changes to https://github.com/plaidml/plaidml/pull/1582

**Pull request closed without merging**